feat(linter): allow fixing in files with source offsets

camchenry opened the pull request:
- fixes https://github.com/oxc-project/oxc/issues/5913

This PR fixes the calculation of spans when dealing with files that have multiple sources and non-zero source start offsets. This is almost always the case for `.vue`, `.astro`, and `.svelte` files. This enables us to correctly apply fixes for these file types both in the CLI with `oxlint` and also in editors like VS Code.


https://github.com/user-attachments/assets/2836c8bd-09be-4e59-801d-7c95f8c2491f



I'm open to ideas on how to improve testing in this area, as I don't think that we currently have any tests for fixing files end-to-end (beyond what the linter rules check). I did run this locally on the gitlab repository (which is written in Vue) and all of the fixes appeared to be applied correctly.

camchenry commented on crates/oxc_linter/src/service.rs:
```diff
@@ -250,26 +251,44 @@ impl Runtime {
                 if fix_result.fixed {
-                    fs::write(path, fix_result.fixed_code.as_bytes()).unwrap();
+                    // write to file, replacing only the changed part
+                    let start = source.start.saturating_add_signed(fix_offset) as usize;
+                    let end = start + source.source_text.len();
+                    new_source_text.to_mut().replace_range(start..end, &fix_result.fixed_code);
+                    let old_code_len = u32::try_from(source.source_text.len()).unwrap();
+                    let new_code_len = u32::try_from(fix_result.fixed_code.len()).unwrap();
+                    fix_offset += i32::try_from(new_code_len).unwrap();
+                    fix_offset -= i32::try_from(old_code_len).unwrap();
```
I would love any ideas on how to avoid all of this number casting here. the main issue is that `start..end` has to be a range of `usize`, but `source.start` is `u32`, while `fix_offset` needs to potentially be negative (as source files can be smaller after applying fixes), so it becomes `i32`.

Boshen replied:
Just cast them with `as`, we assume source length is under u32.

Boshen replied:
https://github.com/oxc-project/oxc/blob/51a78d594640a069db53940bdb838fb420ea698c/crates/oxc_parser/src/lib.rs#L22-L33

camchenry replied:
right, I am often forgetting about that, I'll update this to just cast and ignore truncation

DonIsaac commented:
I like this. I'm currently working on a full re-write of source files and writing, but this is a good in-between. Excellent stuff @camchenry

DonIsaac commented:
I'd like feedback from @camc314. After that, feel free to merge when you're ready

camchenry committed:
feat(linter): allow fixing in files with source offsets (#6197)

- fixes https://github.com/oxc-project/oxc/issues/5913

This PR fixes the calculation of spans when dealing with files that have multiple sources and non-zero source start offsets. This is almost always the case for `.vue`, `.astro`, and `.svelte` files. This enables us to correctly apply fixes for these file types both in the CLI with `oxlint` and also in editors like VS Code.

https://github.com/user-attachments/assets/2836c8bd-09be-4e59-801d-7c95f8c2491f

I'm open to ideas on how to improve testing in this area, as I don't think that we currently have any tests for fixing files end-to-end (beyond what the linter rules check). I did run this locally on the gitlab repository (which is written in Vue) and all of the fixes appeared to be applied correctly.
